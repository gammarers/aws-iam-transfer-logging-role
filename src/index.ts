import { Duration } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

/**
 * @TODO: Not yet supported Omit
 * https://github.com/aws/jsii/issues/4468
 * type omitKeys = 'eventPattern';
 * export interface IamRDSMonitoringRoleProps extends Omit<iam.RoleProps, 'assumedBy'> {}
 */
export interface IamTransferLoggingRoleProps {
  readonly roleName?: string;
  readonly managedPolicies?: iam.IManagedPolicy[];
  readonly permissionsBoundary?: iam.IManagedPolicy;
  readonly maxSessionDuration?: Duration;
  readonly inlinePolicies?: {
    [name: string]: iam.PolicyDocument;
  };
  readonly path?: string;
  readonly description?: string;
}

export class IamTransferLoggingRole extends iam.Role {
  constructor(scope: Construct, id: string, props?: IamTransferLoggingRoleProps) {
    super(scope, id, {
      roleName: props?.roleName,
      permissionsBoundary: props?.permissionsBoundary,
      maxSessionDuration: props?.maxSessionDuration,
      path: props?.path,
      description: (() => {
        return props?.description ?? 'transfer logging role.';
      })(),
      assumedBy: new iam.ServicePrincipal('transfer.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSTransferLoggingAccess'),
      ],
    });
  }
}