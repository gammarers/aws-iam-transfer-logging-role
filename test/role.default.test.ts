import { App, Stack } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import * as iam from 'aws-cdk-lib/aws-iam';
import { IamTransferLoggingRole } from '../src';

describe('IamTransferLoggingRole default Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const role = new IamTransferLoggingRole(stack, 'IamTransferLoggingRole');

  it('Is IAM Role', async () => {
    expect(role).toBeInstanceOf(iam.Role);
  });

  const template = Template.fromStack(stack);

  it('Should match iam role.', async () => {
    template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
      Description: 'transfer logging role.',
      AssumeRolePolicyDocument: Match.objectEquals({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'transfer.amazonaws.com',
            },
          },
        ],
      }),
      ManagedPolicyArns: [
        {
          'Fn::Join': [
            '',
            [
              'arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':iam::aws:policy/service-role/AWSTransferLoggingAccess',
            ],
          ],
        },
      ],
    }));
  });

  it('Should match snapshot.', async () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});