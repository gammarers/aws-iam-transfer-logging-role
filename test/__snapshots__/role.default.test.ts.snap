// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IamTransferLoggingRole default Testing Should match snapshot. 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "IamTransferLoggingRole19259B8A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "transfer.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "transfer logging role.",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSTransferLoggingAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
